{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "244f3951",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3a570ccf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>ACCESSION_NUMBER</th>\n",
       "      <th>GENE_CDS_LENGTH</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>SAMPLE_NAME</th>\n",
       "      <th>ID_SAMPLE</th>\n",
       "      <th>ID_TUMOUR</th>\n",
       "      <th>PRIMARY_SITE</th>\n",
       "      <th>SITE_SUBTYPE_1</th>\n",
       "      <th>SITE_SUBTYPE_2</th>\n",
       "      <th>...</th>\n",
       "      <th>RESISTANCE_MUTATION</th>\n",
       "      <th>MUTATION_SOMATIC_STATUS</th>\n",
       "      <th>PUBMED_PMID</th>\n",
       "      <th>ID_STUDY</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>TUMOUR_ORIGIN</th>\n",
       "      <th>AGE</th>\n",
       "      <th>HGVSP</th>\n",
       "      <th>HGVSC</th>\n",
       "      <th>HGVSG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>2692607</td>\n",
       "      <td>2692607</td>\n",
       "      <td>2551579</td>\n",
       "      <td>thyroid</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>28351340.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surgery-fixed</td>\n",
       "      <td>NS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000418915.1:p.Pro70GlnfsTer49</td>\n",
       "      <td>ENST00000498124.1:c.209_210del</td>\n",
       "      <td>9:g.21971150_21971151del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>902550</td>\n",
       "      <td>902550</td>\n",
       "      <td>820104</td>\n",
       "      <td>pleura</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>12117769.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surgery fresh/frozen</td>\n",
       "      <td>primary</td>\n",
       "      <td>62.0</td>\n",
       "      <td>ENSP00000418915.1:p.His83Gln</td>\n",
       "      <td>ENST00000498124.1:c.249C&gt;A</td>\n",
       "      <td>9:g.21971110G&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>S87946</td>\n",
       "      <td>943503</td>\n",
       "      <td>860138</td>\n",
       "      <td>cervix</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>10408854.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surgery fresh/frozen</td>\n",
       "      <td>NS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000418915.1:p.Gly150=</td>\n",
       "      <td>ENST00000498124.1:c.450T&gt;C</td>\n",
       "      <td>9:g.21970909A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>ME024</td>\n",
       "      <td>1838446</td>\n",
       "      <td>1732547</td>\n",
       "      <td>skin</td>\n",
       "      <td>trunk</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>22817889.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NS</td>\n",
       "      <td>primary</td>\n",
       "      <td>44.0</td>\n",
       "      <td>ENSP00000418915.1:p.Arg87GlyfsTer59</td>\n",
       "      <td>ENST00000498124.1:c.259del</td>\n",
       "      <td>9:g.21971102del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>S88227</td>\n",
       "      <td>943766</td>\n",
       "      <td>860401</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>11753042.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surgery - NOS</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000418915.1:p.Glu61AlafsTer55</td>\n",
       "      <td>ENST00000498124.1:c.177_187del</td>\n",
       "      <td>9:g.21971172_21971182del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3711</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>P-0009960-T01-IM5</td>\n",
       "      <td>2725787</td>\n",
       "      <td>2584608</td>\n",
       "      <td>lung</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>28481359.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surgery-fixed</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000418915.1:p.His123Asn</td>\n",
       "      <td>ENST00000498124.1:c.367C&gt;A</td>\n",
       "      <td>9:g.21970992G&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3712</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>TCGA-85-7699-01</td>\n",
       "      <td>2195194</td>\n",
       "      <td>2063472</td>\n",
       "      <td>lung</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>418.0</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>73.0</td>\n",
       "      <td>ENSP00000418915.1:p.Trp110Ter</td>\n",
       "      <td>ENST00000498124.1:c.329G&gt;A</td>\n",
       "      <td>9:g.21971030C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3713</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>GBM18-I1</td>\n",
       "      <td>2813453</td>\n",
       "      <td>2669845</td>\n",
       "      <td>central_nervous_system</td>\n",
       "      <td>brain</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>28263318.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surgery fresh/frozen</td>\n",
       "      <td>NS</td>\n",
       "      <td>40.0</td>\n",
       "      <td>ENSP00000418915.1:p.Asp84Gly</td>\n",
       "      <td>ENST00000498124.1:c.251A&gt;G</td>\n",
       "      <td>9:g.21971108T&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3714</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>933237</td>\n",
       "      <td>933237</td>\n",
       "      <td>849894</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>ampulla_of_Vater</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>15014024.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surgery-fixed</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000418915.1:p.Glu69Val</td>\n",
       "      <td>ENST00000498124.1:c.206_207delinsTA</td>\n",
       "      <td>9:g.21971152_21971153delinsTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3715</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>1787</td>\n",
       "      <td>1170097</td>\n",
       "      <td>1170097</td>\n",
       "      <td>1082231</td>\n",
       "      <td>upper_aerodigestive_tract</td>\n",
       "      <td>mouth</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>19101490.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surgery fresh/frozen</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000418915.1:p.Leu130Pro</td>\n",
       "      <td>ENST00000498124.1:c.389T&gt;C</td>\n",
       "      <td>9:g.21970970A&gt;G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3716 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENE_NAME   ACCESSION_NUMBER   GENE_CDS_LENGTH   HGNC_ID  \\\n",
       "0       CDKN2A  ENST00000498124.1               504      1787   \n",
       "1       CDKN2A  ENST00000498124.1               504      1787   \n",
       "2       CDKN2A  ENST00000498124.1               504      1787   \n",
       "3       CDKN2A  ENST00000498124.1               504      1787   \n",
       "4       CDKN2A  ENST00000498124.1               504      1787   \n",
       "...        ...                ...               ...       ...   \n",
       "3711    CDKN2A  ENST00000498124.1               504      1787   \n",
       "3712    CDKN2A  ENST00000498124.1               504      1787   \n",
       "3713    CDKN2A  ENST00000498124.1               504      1787   \n",
       "3714    CDKN2A  ENST00000498124.1               504      1787   \n",
       "3715    CDKN2A  ENST00000498124.1               504      1787   \n",
       "\n",
       "            SAMPLE_NAME   ID_SAMPLE   ID_TUMOUR               PRIMARY_SITE  \\\n",
       "0               2692607     2692607     2551579                    thyroid   \n",
       "1                902550      902550      820104                     pleura   \n",
       "2                S87946      943503      860138                     cervix   \n",
       "3                 ME024     1838446     1732547                       skin   \n",
       "4                S88227      943766      860401                   pancreas   \n",
       "...                 ...         ...         ...                        ...   \n",
       "3711  P-0009960-T01-IM5     2725787     2584608                       lung   \n",
       "3712    TCGA-85-7699-01     2195194     2063472                       lung   \n",
       "3713           GBM18-I1     2813453     2669845     central_nervous_system   \n",
       "3714             933237      933237      849894                   pancreas   \n",
       "3715            1170097     1170097     1082231  upper_aerodigestive_tract   \n",
       "\n",
       "        SITE_SUBTYPE_1  SITE_SUBTYPE_2  ...  RESISTANCE_MUTATION  \\\n",
       "0                   NS              NS  ...                    -   \n",
       "1                   NS              NS  ...                    -   \n",
       "2                   NS              NS  ...                    -   \n",
       "3                trunk              NS  ...                    -   \n",
       "4                   NS              NS  ...                    -   \n",
       "...                ...             ...  ...                  ...   \n",
       "3711                NS              NS  ...                    -   \n",
       "3712                NS              NS  ...                    -   \n",
       "3713             brain              NS  ...                    -   \n",
       "3714  ampulla_of_Vater              NS  ...                    -   \n",
       "3715             mouth              NS  ...                    -   \n",
       "\n",
       "        MUTATION_SOMATIC_STATUS  PUBMED_PMID  ID_STUDY           SAMPLE_TYPE  \\\n",
       "0     Variant of unknown origin   28351340.0       NaN         surgery-fixed   \n",
       "1     Variant of unknown origin   12117769.0       NaN  surgery fresh/frozen   \n",
       "2     Variant of unknown origin   10408854.0       NaN  surgery fresh/frozen   \n",
       "3     Variant of unknown origin   22817889.0       NaN                    NS   \n",
       "4     Variant of unknown origin   11753042.0       NaN         surgery - NOS   \n",
       "...                         ...          ...       ...                   ...   \n",
       "3711  Confirmed somatic variant   28481359.0       NaN         surgery-fixed   \n",
       "3712  Confirmed somatic variant          NaN     418.0                    NS   \n",
       "3713  Confirmed somatic variant   28263318.0       NaN  surgery fresh/frozen   \n",
       "3714  Confirmed somatic variant   15014024.0       NaN         surgery-fixed   \n",
       "3715  Confirmed somatic variant   19101490.0       NaN  surgery fresh/frozen   \n",
       "\n",
       "      TUMOUR_ORIGIN   AGE                                HGVSP  \\\n",
       "0                NS   NaN  ENSP00000418915.1:p.Pro70GlnfsTer49   \n",
       "1           primary  62.0         ENSP00000418915.1:p.His83Gln   \n",
       "2                NS   NaN          ENSP00000418915.1:p.Gly150=   \n",
       "3           primary  44.0  ENSP00000418915.1:p.Arg87GlyfsTer59   \n",
       "4           primary   NaN  ENSP00000418915.1:p.Glu61AlafsTer55   \n",
       "...             ...   ...                                  ...   \n",
       "3711        primary   NaN        ENSP00000418915.1:p.His123Asn   \n",
       "3712             NS  73.0        ENSP00000418915.1:p.Trp110Ter   \n",
       "3713             NS  40.0         ENSP00000418915.1:p.Asp84Gly   \n",
       "3714        primary   NaN         ENSP00000418915.1:p.Glu69Val   \n",
       "3715        primary   NaN        ENSP00000418915.1:p.Leu130Pro   \n",
       "\n",
       "                                    HGVSC                          HGVSG  \n",
       "0          ENST00000498124.1:c.209_210del       9:g.21971150_21971151del  \n",
       "1              ENST00000498124.1:c.249C>A                9:g.21971110G>T  \n",
       "2              ENST00000498124.1:c.450T>C                9:g.21970909A>G  \n",
       "3              ENST00000498124.1:c.259del                9:g.21971102del  \n",
       "4          ENST00000498124.1:c.177_187del       9:g.21971172_21971182del  \n",
       "...                                   ...                            ...  \n",
       "3711           ENST00000498124.1:c.367C>A                9:g.21970992G>T  \n",
       "3712           ENST00000498124.1:c.329G>A                9:g.21971030C>T  \n",
       "3713           ENST00000498124.1:c.251A>G                9:g.21971108T>C  \n",
       "3714  ENST00000498124.1:c.206_207delinsTA  9:g.21971152_21971153delinsTA  \n",
       "3715           ENST00000498124.1:c.389T>C                9:g.21970970A>G  \n",
       "\n",
       "[3716 rows x 37 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('CDKN2A main data.csv')\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4392b21b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1386, 37)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filtering for mutation type missense substitution mutation\n",
    "\n",
    "data = data[data[' MUTATION_DESCRIPTION'] == 'Substitution - Missense']\n",
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ce051ef6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1386, 37)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filter to remove mutation points where there is no genomic reference coordinate\n",
    "#this is indicated by value 'null' in 'GRCH' column\n",
    "\n",
    "data = data[data[' GRCH'] != 'null']\n",
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ad3ed589",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this is good, all our records has known substitution information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1d2e0900",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1351, 37)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#now we remove all rows with insertion, duplication, inversion or deletion information\n",
    "\n",
    "data = data[~data[' HGVSG'].str.contains('del|in|dup', case=False, na=False, regex=True)]\n",
    "data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c7c5bb6",
   "metadata": {},
   "source": [
    "We are done with all the filtering we need. We will now arrange the dataframe to better do our analyses. First, we build out our genomic level features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e30b39dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns_to_copy = ['GENE_NAME', ' ACCESSION_NUMBER', ' GENE_CDS_LENGTH', ' HGVSG', ' HGVSC', ' MUTATION_CDS']\n",
    "new_df = data[columns_to_copy].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bcd9eda4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract position of mutation\n",
    "new_df['CDNA_POSITION'] = new_df[' MUTATION_CDS'].str.extract('(\\d+)')\n",
    "new_df['CDNA_POSITION'] = pd.to_numeric(new_df['CDNA_POSITION'], errors='coerce')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9004a32b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract mutation event, base and mutant alleles\n",
    "new_df['MUTATION_EVENT'] = new_df[' MUTATION_CDS'].str[-3:]\n",
    "new_df['BASE_ALLELE'] = new_df['MUTATION_EVENT'].str[0]\n",
    "new_df['MUTANT_ALLELE'] = new_df['MUTATION_EVENT'].str[-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "409ae34d",
   "metadata": {},
   "source": [
    "Now we get our proteomic level features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e816b5a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_mapping = {\n",
    "    'A': 'Ala',\n",
    "    'R': 'Arg',\n",
    "    'N': 'Asn',\n",
    "    'D': 'Asp',\n",
    "    'C': 'Cys',\n",
    "    'Q': 'Gln',\n",
    "    'E': 'Glu',\n",
    "    'G': 'Gly',\n",
    "    'H': 'His',\n",
    "    'I': 'Ile',\n",
    "    'L': 'Leu',\n",
    "    'K': 'Lys',\n",
    "    'M': 'Met',\n",
    "    'F': 'Phe',\n",
    "    'P': 'Pro',\n",
    "    'S': 'Ser',\n",
    "    'T': 'Thr',\n",
    "    'W': 'Trp',\n",
    "    'Y': 'Tyr',\n",
    "    'V': 'Val'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "efd7e5b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df[['HGVSP', 'MUTATION_AA']] = data[[' HGVSP', ' MUTATION_AA']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b330a79e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract wild-type and mutant amino acids and mutation position in 1-letter and 3-letter format\n",
    "new_df['WT_AA_1'] = new_df['MUTATION_AA'].str[2]\n",
    "new_df['WT_AA_3'] = new_df['WT_AA_1'].map(aa_mapping)\n",
    "\n",
    "new_df['CODON_POSITION'] = new_df['MUTATION_AA'].str.extract('(\\d+)')\n",
    "new_df['CODON_POSITION'] = pd.to_numeric(new_df['CODON_POSITION'], errors='coerce')\n",
    "\n",
    "new_df['MT_AA_1'] = new_df['MUTATION_AA'].str[-1]\n",
    "new_df['MT_AA_3'] = new_df['MT_AA_1'].map(aa_mapping)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf090ad0",
   "metadata": {},
   "source": [
    "Add all your site and histology specific features now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7976fb82",
   "metadata": {},
   "outputs": [],
   "source": [
    " new_df[['MUTATION_SOMATIC_STATUS', 'PRIMARY_SITE', 'PRIMARY_HISTOLOGY', 'SAMPLE_TYPE', 'TUMOUR_ORIGIN', 'GENOMIC_MUTATION_ID', 'PUBMED_PMID']] = data[[' MUTATION_SOMATIC_STATUS', ' PRIMARY_SITE', ' PRIMARY_HISTOLOGY', ' SAMPLE_TYPE', ' TUMOUR_ORIGIN', ' GENOMIC_MUTATION_ID', ' PUBMED_PMID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ee5e97ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df['PRIMARY_SITE'] = new_df['PRIMARY_SITE'].str.replace('_', ' ')\n",
    "new_df['PRIMARY_HISTOLOGY'] = new_df['PRIMARY_HISTOLOGY'].str.replace('_', ' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0c57fa1e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>ACCESSION_NUMBER</th>\n",
       "      <th>GENE_CDS_LENGTH</th>\n",
       "      <th>HGVSG</th>\n",
       "      <th>HGVSC</th>\n",
       "      <th>MUTATION_CDS</th>\n",
       "      <th>CDNA_POSITION</th>\n",
       "      <th>MUTATION_EVENT</th>\n",
       "      <th>BASE_ALLELE</th>\n",
       "      <th>MUTANT_ALLELE</th>\n",
       "      <th>...</th>\n",
       "      <th>CODON_POSITION</th>\n",
       "      <th>MT_AA_1</th>\n",
       "      <th>MT_AA_3</th>\n",
       "      <th>MUTATION_SOMATIC_STATUS</th>\n",
       "      <th>PRIMARY_SITE</th>\n",
       "      <th>PRIMARY_HISTOLOGY</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>TUMOUR_ORIGIN</th>\n",
       "      <th>GENOMIC_MUTATION_ID</th>\n",
       "      <th>PUBMED_PMID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21971110G&gt;T</td>\n",
       "      <td>ENST00000498124.1:c.249C&gt;A</td>\n",
       "      <td>c.249C&gt;A</td>\n",
       "      <td>249</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>83</td>\n",
       "      <td>Q</td>\n",
       "      <td>Gln</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>pleura</td>\n",
       "      <td>mesothelioma</td>\n",
       "      <td>surgery fresh/frozen</td>\n",
       "      <td>primary</td>\n",
       "      <td>COSV58684711</td>\n",
       "      <td>12117769.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21971171A&gt;G</td>\n",
       "      <td>ENST00000498124.1:c.188T&gt;C</td>\n",
       "      <td>c.188T&gt;C</td>\n",
       "      <td>188</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>63</td>\n",
       "      <td>P</td>\n",
       "      <td>Pro</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>skin</td>\n",
       "      <td>malignant melanoma</td>\n",
       "      <td>surgery - NOS</td>\n",
       "      <td>metastasis</td>\n",
       "      <td>COSV58716983</td>\n",
       "      <td>20703244.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21971135G&gt;A</td>\n",
       "      <td>ENST00000498124.1:c.224C&gt;T</td>\n",
       "      <td>c.224C&gt;T</td>\n",
       "      <td>224</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>75</td>\n",
       "      <td>L</td>\n",
       "      <td>Leu</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>lung</td>\n",
       "      <td>carcinoma</td>\n",
       "      <td>surgery-fixed</td>\n",
       "      <td>metastasis</td>\n",
       "      <td>COSV58688496</td>\n",
       "      <td>26164066.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21974742C&gt;T</td>\n",
       "      <td>ENST00000498124.1:c.86G&gt;A</td>\n",
       "      <td>c.86G&gt;A</td>\n",
       "      <td>86</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>29</td>\n",
       "      <td>Q</td>\n",
       "      <td>Gln</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>upper aerodigestive tract</td>\n",
       "      <td>carcinoma</td>\n",
       "      <td>surgery-fixed</td>\n",
       "      <td>NS</td>\n",
       "      <td>COSV58704956</td>\n",
       "      <td>19690981.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21971203C&gt;G</td>\n",
       "      <td>ENST00000498124.1:c.156G&gt;C</td>\n",
       "      <td>c.156G&gt;C</td>\n",
       "      <td>156</td>\n",
       "      <td>G&gt;C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>52</td>\n",
       "      <td>I</td>\n",
       "      <td>Ile</td>\n",
       "      <td>Variant of unknown origin</td>\n",
       "      <td>breast</td>\n",
       "      <td>carcinoma</td>\n",
       "      <td>cell-line</td>\n",
       "      <td>NS</td>\n",
       "      <td>COSV58687020</td>\n",
       "      <td>19593635.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3700</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21971117G&gt;A</td>\n",
       "      <td>ENST00000498124.1:c.242C&gt;T</td>\n",
       "      <td>c.242C&gt;T</td>\n",
       "      <td>242</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>L</td>\n",
       "      <td>Leu</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>skin</td>\n",
       "      <td>malignant melanoma</td>\n",
       "      <td>surgery-fixed</td>\n",
       "      <td>metastasis</td>\n",
       "      <td>COSV58683884</td>\n",
       "      <td>28481359.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3706</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21971109C&gt;T</td>\n",
       "      <td>ENST00000498124.1:c.250G&gt;A</td>\n",
       "      <td>c.250G&gt;A</td>\n",
       "      <td>250</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>84</td>\n",
       "      <td>N</td>\n",
       "      <td>Asn</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>central nervous system</td>\n",
       "      <td>primitive neuroectodermal tumour-medulloblastoma</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>COSV58683289</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3711</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21970992G&gt;T</td>\n",
       "      <td>ENST00000498124.1:c.367C&gt;A</td>\n",
       "      <td>c.367C&gt;A</td>\n",
       "      <td>367</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>123</td>\n",
       "      <td>N</td>\n",
       "      <td>Asn</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>lung</td>\n",
       "      <td>carcinoma</td>\n",
       "      <td>surgery-fixed</td>\n",
       "      <td>primary</td>\n",
       "      <td>COSV58709909</td>\n",
       "      <td>28481359.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3713</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21971108T&gt;C</td>\n",
       "      <td>ENST00000498124.1:c.251A&gt;G</td>\n",
       "      <td>c.251A&gt;G</td>\n",
       "      <td>251</td>\n",
       "      <td>A&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>84</td>\n",
       "      <td>G</td>\n",
       "      <td>Gly</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>central nervous system</td>\n",
       "      <td>glioma</td>\n",
       "      <td>surgery fresh/frozen</td>\n",
       "      <td>NS</td>\n",
       "      <td>COSV58688961</td>\n",
       "      <td>28263318.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3715</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>ENST00000498124.1</td>\n",
       "      <td>504</td>\n",
       "      <td>9:g.21970970A&gt;G</td>\n",
       "      <td>ENST00000498124.1:c.389T&gt;C</td>\n",
       "      <td>c.389T&gt;C</td>\n",
       "      <td>389</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>130</td>\n",
       "      <td>P</td>\n",
       "      <td>Pro</td>\n",
       "      <td>Confirmed somatic variant</td>\n",
       "      <td>upper aerodigestive tract</td>\n",
       "      <td>carcinoma</td>\n",
       "      <td>surgery fresh/frozen</td>\n",
       "      <td>primary</td>\n",
       "      <td>COSV58702833</td>\n",
       "      <td>19101490.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1351 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENE_NAME   ACCESSION_NUMBER   GENE_CDS_LENGTH            HGVSG  \\\n",
       "1       CDKN2A  ENST00000498124.1               504  9:g.21971110G>T   \n",
       "16      CDKN2A  ENST00000498124.1               504  9:g.21971171A>G   \n",
       "17      CDKN2A  ENST00000498124.1               504  9:g.21971135G>A   \n",
       "18      CDKN2A  ENST00000498124.1               504  9:g.21974742C>T   \n",
       "20      CDKN2A  ENST00000498124.1               504  9:g.21971203C>G   \n",
       "...        ...                ...               ...              ...   \n",
       "3700    CDKN2A  ENST00000498124.1               504  9:g.21971117G>A   \n",
       "3706    CDKN2A  ENST00000498124.1               504  9:g.21971109C>T   \n",
       "3711    CDKN2A  ENST00000498124.1               504  9:g.21970992G>T   \n",
       "3713    CDKN2A  ENST00000498124.1               504  9:g.21971108T>C   \n",
       "3715    CDKN2A  ENST00000498124.1               504  9:g.21970970A>G   \n",
       "\n",
       "                           HGVSC  MUTATION_CDS  CDNA_POSITION MUTATION_EVENT  \\\n",
       "1     ENST00000498124.1:c.249C>A      c.249C>A            249            C>A   \n",
       "16    ENST00000498124.1:c.188T>C      c.188T>C            188            T>C   \n",
       "17    ENST00000498124.1:c.224C>T      c.224C>T            224            C>T   \n",
       "18     ENST00000498124.1:c.86G>A       c.86G>A             86            G>A   \n",
       "20    ENST00000498124.1:c.156G>C      c.156G>C            156            G>C   \n",
       "...                          ...           ...            ...            ...   \n",
       "3700  ENST00000498124.1:c.242C>T      c.242C>T            242            C>T   \n",
       "3706  ENST00000498124.1:c.250G>A      c.250G>A            250            G>A   \n",
       "3711  ENST00000498124.1:c.367C>A      c.367C>A            367            C>A   \n",
       "3713  ENST00000498124.1:c.251A>G      c.251A>G            251            A>G   \n",
       "3715  ENST00000498124.1:c.389T>C      c.389T>C            389            T>C   \n",
       "\n",
       "     BASE_ALLELE MUTANT_ALLELE  ... CODON_POSITION MT_AA_1 MT_AA_3  \\\n",
       "1              C             A  ...             83       Q     Gln   \n",
       "16             T             C  ...             63       P     Pro   \n",
       "17             C             T  ...             75       L     Leu   \n",
       "18             G             A  ...             29       Q     Gln   \n",
       "20             G             C  ...             52       I     Ile   \n",
       "...          ...           ...  ...            ...     ...     ...   \n",
       "3700           C             T  ...             81       L     Leu   \n",
       "3706           G             A  ...             84       N     Asn   \n",
       "3711           C             A  ...            123       N     Asn   \n",
       "3713           A             G  ...             84       G     Gly   \n",
       "3715           T             C  ...            130       P     Pro   \n",
       "\n",
       "        MUTATION_SOMATIC_STATUS               PRIMARY_SITE  \\\n",
       "1     Variant of unknown origin                     pleura   \n",
       "16    Variant of unknown origin                       skin   \n",
       "17    Variant of unknown origin                       lung   \n",
       "18    Variant of unknown origin  upper aerodigestive tract   \n",
       "20    Variant of unknown origin                     breast   \n",
       "...                         ...                        ...   \n",
       "3700  Confirmed somatic variant                       skin   \n",
       "3706  Confirmed somatic variant     central nervous system   \n",
       "3711  Confirmed somatic variant                       lung   \n",
       "3713  Confirmed somatic variant     central nervous system   \n",
       "3715  Confirmed somatic variant  upper aerodigestive tract   \n",
       "\n",
       "                                     PRIMARY_HISTOLOGY           SAMPLE_TYPE  \\\n",
       "1                                         mesothelioma  surgery fresh/frozen   \n",
       "16                                  malignant melanoma         surgery - NOS   \n",
       "17                                           carcinoma         surgery-fixed   \n",
       "18                                           carcinoma         surgery-fixed   \n",
       "20                                           carcinoma             cell-line   \n",
       "...                                                ...                   ...   \n",
       "3700                                malignant melanoma         surgery-fixed   \n",
       "3706  primitive neuroectodermal tumour-medulloblastoma                    NS   \n",
       "3711                                         carcinoma         surgery-fixed   \n",
       "3713                                            glioma  surgery fresh/frozen   \n",
       "3715                                         carcinoma  surgery fresh/frozen   \n",
       "\n",
       "     TUMOUR_ORIGIN GENOMIC_MUTATION_ID PUBMED_PMID  \n",
       "1          primary        COSV58684711  12117769.0  \n",
       "16      metastasis        COSV58716983  20703244.0  \n",
       "17      metastasis        COSV58688496  26164066.0  \n",
       "18              NS        COSV58704956  19690981.0  \n",
       "20              NS        COSV58687020  19593635.0  \n",
       "...            ...                 ...         ...  \n",
       "3700    metastasis        COSV58683884  28481359.0  \n",
       "3706            NS        COSV58683289         NaN  \n",
       "3711       primary        COSV58709909  28481359.0  \n",
       "3713            NS        COSV58688961  28263318.0  \n",
       "3715       primary        COSV58702833  19101490.0  \n",
       "\n",
       "[1351 rows x 24 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "aed02743",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts of each distinct value in PRIMARY_HISTOLOGY column:\n",
      "carcinoma                                              982\n",
      "malignant melanoma                                     163\n",
      "other                                                   39\n",
      "lymphoid neoplasm                                       37\n",
      "glioma                                                  35\n",
      "haematopoietic neoplasm                                 24\n",
      "adnexal tumour                                          20\n",
      "meningioma                                               7\n",
      "gastrointestinal stromal tumour                          3\n",
      "germ cell tumour                                         3\n",
      "carcinoid-endocrine tumour                               3\n",
      "benign melanocytic nevus                                 3\n",
      "primitive neuroectodermal tumour-medulloblastoma         2\n",
      "mesothelioma                                             2\n",
      "pancreatic intraepithelial neoplasia (PanIN)             2\n",
      "sarcoma                                                  2\n",
      "malignant fibrous histiocytoma-pleomorphic sarcoma       2\n",
      "neuroblastoma                                            2\n",
      "adrenal cortical carcinoma                               2\n",
      "pheochromocytoma                                         2\n",
      "fibrosarcoma                                             2\n",
      "thymic carcinoma                                         1\n",
      "calcifying epithelial odontogenic tumour                 1\n",
      "Wilms tumour                                             1\n",
      "adenoma                                                  1\n",
      "malignant melanoma of soft parts-clear cell sarcoma      1\n",
      "fibroepithelial neoplasm                                 1\n",
      "in situ neoplasm                                         1\n",
      "rhabdomyosarcoma                                         1\n",
      "leiomyosarcoma                                           1\n",
      "NS                                                       1\n",
      "paraganglioma                                            1\n",
      "chondrosarcoma                                           1\n",
      "in situ epithelial neoplasm                              1\n",
      "low malignant potential (borderline) tumour              1\n",
      "Name: PRIMARY_HISTOLOGY, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# distinct_primary_histology_values = new_df['PRIMARY_HISTOLOGY'].unique()\n",
    "\n",
    "# print(\"Distinct Values in PRIMARY_HISTOLOGY column:\")\n",
    "# print()\n",
    "# for value in distinct_primary_histology_values:\n",
    "#     print(value)\n",
    "    \n",
    "    \n",
    "distinct_primary_histology_values = new_df['PRIMARY_HISTOLOGY'].unique()\n",
    "\n",
    "print(\"Counts of each distinct value in PRIMARY_HISTOLOGY column:\")\n",
    "print(new_df['PRIMARY_HISTOLOGY'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "86ae5748",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_type_mappings = {\n",
    "    'mesothelioma': 'Other',\n",
    "    'malignant melanoma': 'Skin Cancer',\n",
    "    'carcinoma': 'Carcinoma',\n",
    "    'lymphoid neoplasm': 'Lymphoma',\n",
    "    'haematopoietic neoplasm': 'Leukemia',\n",
    "    'adnexal tumour': 'Skin Cancer',\n",
    "    'glioma': 'Brain Tumor',\n",
    "    'gastrointestinal stromal tumour': 'Other',\n",
    "    'fibrosarcoma': 'Sarcoma',\n",
    "    'rhabdomyosarcoma': 'Sarcoma',\n",
    "    'other': 'Other',\n",
    "    'carcinoid-endocrine tumour': 'Other',\n",
    "    'pheochromocytoma': 'Other',\n",
    "    'meningioma': 'Brain Tumor',\n",
    "    'malignant melanoma of soft parts-clear cell sarcoma': 'Sarcoma',\n",
    "    'Wilms tumour': 'Other',\n",
    "    'calcifying epithelial odontogenic tumour': 'Other',\n",
    "    'low malignant potential (borderline) tumour': 'Other',\n",
    "    'thymic carcinoma': 'Other',\n",
    "    'sarcoma': 'Sarcoma',\n",
    "    'primitive neuroectodermal tumour-medulloblastoma': 'Brain Tumor',\n",
    "    'germ cell tumour': 'Other',\n",
    "    'adenoma': 'Other',\n",
    "    'benign melanocytic nevus': 'Skin Cancer',\n",
    "    'fibroepithelial neoplasm': 'Other',\n",
    "    'pancreatic intraepithelial neoplasia (PanIN)': 'Other',\n",
    "    'in situ neoplasm': 'Other',\n",
    "    'neuroblastoma': 'Brain Tumor',\n",
    "    'leiomyosarcoma': 'Sarcoma',\n",
    "    'NS': 'Other',\n",
    "    'paraganglioma': 'Other',\n",
    "    'malignant fibrous histiocytoma-pleomorphic sarcoma': 'Sarcoma',\n",
    "    'chondrosarcoma': 'Other',\n",
    "    'in situ epithelial neoplasm': 'Carcinoma',\n",
    "    'adrenal cortical carcinoma': 'Other'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "79dd1803",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     PRIMARY_HISTOLOGY  CANCER_TYPE\n",
      "1                                         mesothelioma        Other\n",
      "16                                  malignant melanoma  Skin Cancer\n",
      "17                                           carcinoma    Carcinoma\n",
      "18                                           carcinoma    Carcinoma\n",
      "20                                           carcinoma    Carcinoma\n",
      "...                                                ...          ...\n",
      "3700                                malignant melanoma  Skin Cancer\n",
      "3706  primitive neuroectodermal tumour-medulloblastoma  Brain Tumor\n",
      "3711                                         carcinoma    Carcinoma\n",
      "3713                                            glioma  Brain Tumor\n",
      "3715                                         carcinoma    Carcinoma\n",
      "\n",
      "[1351 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "new_df['CANCER_TYPE'] = 'Other'  \n",
    "new_df['CANCER_TYPE'] = new_df['PRIMARY_HISTOLOGY'].map(cancer_type_mappings)\n",
    "print(new_df[['PRIMARY_HISTOLOGY', 'CANCER_TYPE']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bf018139",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distinct Values in CANCER_TYPE column:\n",
      "Carcinoma      983\n",
      "Skin Cancer    186\n",
      "Other           66\n",
      "Brain Tumor     46\n",
      "Lymphoma        37\n",
      "Leukemia        24\n",
      "Sarcoma          9\n",
      "Name: CANCER_TYPE, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "distinct_cancers = new_df['CANCER_TYPE'].unique()\n",
    "\n",
    "print(\"Distinct Values in CANCER_TYPE column:\")\n",
    "print(new_df['CANCER_TYPE'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "459dfe95",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df.to_csv('filtered_df.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53dd97d2",
   "metadata": {},
   "source": [
    "Now we have our filtered file which we can use for analysis."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
